governance/status.go

```diff
@@ -0,0 +1,305 @@
+	LastUpdated   string               `json:"lastUpdated"`
+}
+
+func (w DeviceStatus) String() string {
+	return fmt.Sprintf(
+		"Connectivity: %v, "+
+			"Microservices: %v, "+
+			"Workloads: %v,"+
+			"LastUpdated: %v",
+		w.Connectivity, w.Microservices, w.Workloads, w.LastUpdated)
+}
+
+func NewDeviceStatus() *DeviceStatus {
+	var ds DeviceStatus
+	ds.Connectivity = make(map[string]bool, 0)
+	ds.Microservices = make([]MicroserviceStatus, 0)
+	ds.Workloads = make([]WorkloadStatus, 0)
+	return &ds
+}
+
+// Report the containers status and connectivity status to the exchange.
+func (w *GovernanceWorker) ReportDeviceStatus() {
+
+	if !w.Config.Edge.ReportDeviceStatus {
+		glog.Info("ReportDeviceStatus is false. The status report to the exchange is turned off.")
+		return
+	}
+
+	glog.Info("started the status report to the exchange.")
+
+	w.deviceStatus = nil
+	var device_status DeviceStatus
+
+	// get connectivity
+	connect := make(map[string]bool, 0)
+	for _, host := range HORIZON_SERVERS {
+		if err := cutil.CheckConnectivity(host); err != nil {
+			glog.Errorf(logString(fmt.Sprintf("Error checking connectivity for %s: %v", host, err)))
+			connect[host] = false
+		} else {
+			connect[host] = true
+		}
+	}
+	device_status.Connectivity = connect
+
+	// get docker containers
+	containers := make([]docker.APIContainers, 0)
+	if client, err := docker.NewClient(w.Config.Edge.DockerEndpoint); err != nil {
+		glog.Errorf(logString(fmt.Sprintf("Failed to instantiate docker Client: %v", err)))
+	} else {
+		containers, err = client.ListContainers(docker.ListContainersOptions{})
+		if err != nil {
+			glog.Errorf(logString(fmt.Sprintf("Unable to get list of running containers: %v", err)))
+		}
+	}
+
+	// get microservice status
+	if ms_status, err := w.getMicroserviceStatus(containers); err != nil {
+		glog.Errorf(logString(fmt.Sprintf("Error getting microservice container status: %v", err)))
+	} else {
+		device_status.Microservices = ms_status
+	}
+
+	// get workload status
+	if wl_status, err := w.getWorkloadStatus(containers); err != nil {
+		glog.Errorf(logString(fmt.Sprintf("Error getting microservice container status: %v", err)))
+	} else {
+		device_status.Workloads = wl_status
+	}
+
+	//compare the device status with the one gotten from last time
+	if jbytes, err := json.Marshal(&device_status); err != nil {
+		glog.V(5).Infof(logString(fmt.Sprintf("Failed to convert the device status %v to json: %v", device_status, err)))
+	} else {
+
+		w.deviceStatus = &device_status
+		glog.V(5).Infof(logString(fmt.Sprintf("device status to report to the exchange: %v", string(jbytes[:]))))
+
+		// report the status to the exchange
+		if err := w.writeStatusToExchange(&device_status); err != nil {
+			glog.Errorf(logString(err))
+		}
+	}
+
+}
+
+// Find the status for all the microservices
+func (w *GovernanceWorker) getMicroserviceStatus(containers []docker.APIContainers) ([]MicroserviceStatus, error) {
```

Comment:
We need unit tests for getMicroserviceStatus and getWorkloadStatus just like you have for getContainerStatus.

Reply:
We decided to defer those to an issue, Ling is opening it.